### ISSUE TITLE
Incorrect account-address-decoder causes wallet to crash

### ISSUE THREAD

**Issue by Anviking** [cardano-foundation/cardano-wallet#631]
| Release | Operating System       | Cause |
| ---     | ---                    | ---   |
| next    | Windows & OSX & Linux) | Code  v Configuration v Environment v Human v Unknown |

# Context

There are multiple types of addresses. I believe we have been only using and testing single-addresses.

https://github.com/input-output-hk/cardano-wallet/blob/a7e23556f29923cf9a30794d40a22e8dad969fd9/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs#L509-L512
 
 

- https://input-output-rnd.slack.com/archives/C819S481Y/p1565716791001600
- https://input-output-rnd.slack.com/archives/C819S481Y/p1565868975017100
# Steps to Reproduce

1. Create an account address, e.g. `ca1s5xqmu4e5nxnarxndwqmu2277244hgje9x3kxk0cfng65h8tutaw6za70m9` using `jcli address account --testing ed25519_pk1psxl9wdye5lge5mtsxlzjhhj4dd6ykff5d34n7zv6x49e6lzltksr08tvz
ca1s5xqmu4e5nxnarxndwqmu2277244hgje9x3kxk0cfng65h8tutaw6za70m9` 
2. Use the address in the genesis.yaml of the integration tests:
```diff
diff --git a/lib/jormungandr/test/data/jormungandr/genesis.yaml b/lib/jormungandr/test/data/jormungandr/genesis.yaml
index 9bee6ab9..de3aaa89 100644
--- a/lib/jormungandr/test/data/jormungandr/genesis.yaml
+++ b/lib/jormungandr/test/data/jormungandr/genesis.yaml
@@ -63,7 +63,7 @@ blockchain_configuration:
 # address.
 initial:
   - fund:
-    - address: ta1swk7svu8avn5jysl83vja72lp0sqfczanqee6q08a4j7q27a77kpg4p254a
+    - address: ca1s5xqmu4e5nxnarxndwqmu2277244hgje9x3kxk0cfng65h8tutaw6za70m9
       value: 100000000000
   - fund:
     - address: ta1sv3jh4rlgz3lh8yyyj5cr4qdewewpdrusexfx6x6tzt3syzy2w4d5urjsly
```
3. Re-generate the `block0` binary `jcli genesis encode --input lib/jormungandr/test/data/jormungandr/genesis.yaml --output lib/jormungandr/test/data/jormungandr/block0.bin`
4. You might have to run `rm /tmp/cardano-wallet-jormungandr/storage/blocks.sqlite`
5. Run `jormungandr --genesis-block lib/jormungandr/test/data/jormungandr/block0.bin --config lib/jormungandr/test/data/jormungandr/config.yaml --secret lib/jormungandr/test/data/jormungandr/secret.yaml`
6. In a separate terminal, run `stack exec -- cardano-wallet-jormungandr serve --genesis-hash (jcli genesis hash --input lib/jormungandr/test/data/jormungandr/block0.bin)`, or `$(` instead of `(` with bash.

## Expected behavior

1. Wallet doesn't crash when decoding a valid account address
2. Wallet doesn't crash when decoding *any* valid address

## Actual behavior
 
<A clear and concise description of what you observe instead. If applicable add screenshots to help explain your problem.>

Wallet crashes.

```
stack exec -- cardano-wallet-jormungandr serve --genesis-hash (jcli genesis hash --input lib/jormungandr/test/data/jormungandr/block0.bin)
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-08-15 13:04:18.69 UTC] Wallet backend server starting...
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-08-15 13:04:18.69 UTC] Running as v2019.7.24
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-08-15 13:04:18.69 UTC] Node is Jörmungandr on testnet
[iohk.cardano-wallet.serve.database:Notice:ThreadId 4] [2019-08-15 13:04:18.70 UTC] 10 migrations were applied to the database.
[iohk.cardano-wallet.serve:Info:ThreadId 4] [2019-08-15 13:04:18.70 UTC] Waiting for Jörmungandr to be ready on tcp/8080
cardano-wallet-jormungandr: Data.Binary.Get.runGet at position 213: isolate: the decoder consumed 129 bytes which is less than the expected 46129 bytes
getBlock
CallStack (from HasCallStack):
  error, called at libraries/binary/src/Data/Binary/Get.hs:351:5 in binary-0.8.6.0:Data.Binary.Get
```

---

# Resolution Plan

- We should probably improve our testing coverage to different types of addresses.
- The account-address decoder used an incorrect length. Here is a fix:
```diff
diff --git a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs
index 2cffa813..192c7497 100644
--- a/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs
+++ b/lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs
@@ -508,7 +508,7 @@ getAddress = do
     case kind of
         0x3 -> Address <$> getByteString 33 -- single address
         0x4 -> Address <$> getByteString 65 -- grouped address
-        0x5 -> Address <$> getByteString 65 -- account address
+        0x5 -> Address <$> getByteString 33 -- account address
         0x6 -> Address <$> getByteString 33 -- multisig address
         other -> fail $ "Invalid address type: " ++ show other
   where
```

# PR

| Number   | Base      |
| ---      | ---       |
| #632        | `master` |
| #633        | `master` |

# QA 

- There is a golden test, see PR #632.
- Steps-to-reproduce should work as expected now too
- Next steps <s>(but outside this ticket):</s> better ensure that the binary format is correct by testing against more data from `jormungandr`/`jcli`. We make sure we can deserialise arbitrary genesis-blocks generated by `jormungandr-lib` (looks like this https://github.com/input-output-hk/cardano-wallet/files/3517383/block0-1.txt). Contains `Single`, `Grouped`, and `Account`-addresses. Multisig addresses don't yet exist in jormungandr.

**Comment by piotr-iohk** [comment]
Just a few questions:
1. so, I'm able to go through _Steps to reproduce_ and it works successfully, i.e. I can start `jormungandr`, but I'm not able to run integration tests on this regenerated block0, immediatelly getting:

```
  src/Test/Integration/Scenario/API/Transactions.hs:1028:5: 
  1) PR_DISABLED Transactions API endpoint tests TRANS_LIST_02 - Start time shouldn't be later than end time
       uncaught exception: IOException of type UserError
       user error (unexpected error: ErrGetBlockchainParamsGenesisNotFound (Hash {getHash = "\219\165\151\190\229\240\152~\251\245ok\215\244L8\NAK\138wp\208\203(\162k^\202@a:~\189"}))
```
is there something missing? I thought since I can start `jormungandr` I should be able to run tests too...


2. there is added a golden test for `account` address, maybe that's good opportunity to add also such tests for single, grouped and multisig (what's that btw?).

3. Looking at the linked _slacks_ - interesting case pointed here actually -> https://input-output-rnd.slack.com/archives/CCPKJ93Q9/p1565888697055800. I suppose `./test/data/jormungandr/genesis.yaml` could be modifued to include it (e.g. remove `- fund: ` from the first few transactions, if I understand correctly, our integration tests should work anyway). Unfortunatelly, again, I see the same error after modifing `genesis.yaml` and regenerating `block0` as in `1.` when trying to run integration tests...

**Comment by Anviking** [comment]
> there is added a golden test for account address, maybe that's good opportunity to add also such tests for single, grouped and multisig (what's that btw?).

Good points. When thinking about it, that should have been covered by #633. I checked the printed rust-objects again (actually, I maybe should add that functionality to the patch), and it seems that method only produces account- and and single-addresses. [block0-1.txt](https://github.com/input-output-hk/cardano-wallet/files/3517383/block0-1.txt)

[Jörmungandr addresses](https://github.com/input-output-hk/chain-libs/blob/a3aed5bf07be514dc3c8f4209f3cca312a5f1ca5/chain-addr/src/lib.rs#L10-L14)

My understanding: We are already using (and testing deserialization of) single addresses. Grouped are related to staking (Are they going to be the default then?).  Account addresses are for account-accounting (instead of UTxO), which we won't support [anytime soon](https://input-output-rnd.slack.com/archives/C819S481Y/p1565936332020400?thread_ts=1565868975.017100&cid=C819S481Y). Multisig-addresses require more than one key/signature to be spent.

> so, I'm able to go through Steps to reproduce and it works successfully, i.e. I can start jormungandr, but I'm not able to run integration tests on this regenerated block0, immediatelly getting:

I have not replicated this, but I bet it is due to the genesis-hash having changed, while the integration tests use a [hard-coded one](https://github.com/input-output-hk/cardano-wallet/blob/5be0631d3341f0f4a9fed56f9c2b0e7a9f202c76/lib/jormungandr/test/integration/Main.hs#L199).

That aside, I'm not sure the integration tests will run still because with the patch we are no longer putting funds in the `ta1swk7svu8avn5jysl83vja72lp0sqfczanqee6q08a4j7q27a77kpg4p254a` address, which our faucet won't expect.

I'll check my sanity tomorrow. Purely adding another `-fund` line for an account address should work.

**Comment by Anviking** [comment]
> I'll check my sanity tomorrow. Purely adding another -fund line for an account address should work.

Yes this worked when I tried it (but also very expected).

The "Steps to reproduce" weren't written to work with the integration tests. Given that launching a wallet manually works, I _think_ it's safe to say that any integration test issues are likely unrelated, or at the very least out-of-scope.

**Comment by Anviking** [comment]
Patch that only appends a normal single address to the genesis block:
[update-genesis-block-patch.txt](https://github.com/input-output-hk/cardano-wallet/files/3534842/update-genesis-block-patch.txt)

[Output (don't remember if it was with or without the minor fix below)](https://github.com/input-output-hk/cardano-wallet/files/3534889/jormungandr-integration-output.txt)


With it I get similar errors. Not sure why.

Updating an additional occurrence of the genesis hash in `jormungandr/test/integration/Test/Integration/Jormungandr/Scenario/CLI/Server.hs` only helped a little bit.

**Comment by piotr-iohk** [comment]
lgtm.

I have addressed point 3 in #661
